### DISCUSSION TITLE
SAGEConv and Skip

### DISCUSSION THREAD

**Discussion by DuanhaoranCC** [pyg-team/pytorch_geometric#4613] [General]
I want to ask why graphsage does not support GCN aggregation. In addition, I would like to ask how to realize jump connection and whether my following implementation methods are correct. Thank you very much for your answer!
```python
from torch_geometric.nn import GCNConv, SAGEConv

class Net(torch.nn.Module):
    def __init__(self, p2, p3):
        super().__init__()
        self.conv1 = SAGEConv(train_dataset.num_features, 256, aggr='mean')
        self.lin1 = torch.nn.Linear(train_dataset.num_features, 256)
        self.conv2 = SAGEConv(256, 256, aggr='mean')
        self.conv3 = SAGEConv(256, 256, aggr='mean')
        self.dp2 = nn.Dropout(p=p2)
        self.dp3 = nn.Dropout(p=p3)
        self.action = nn.PReLU()

    def forward(self, x, edge_index):
        h1 = self.action(self.conv1(x, edge_index))
        h1 = self.dp2(h1)
        h2 = self.action(self.conv2(h1 + self.lin1(x), edge_index))
        h2 = self.dp3(h2)
        h3 = self.action(self.conv3(self.lin1(x) + h2 + h1, edge_index))
        return h3
```

**Comment by rusty1s** [comment]
By GCN aggregation you mean symmetric normalization? Does the GraphSAGE paper actually proposes to do that? Your skip-connection implementation looks good to me.

**Reply by DuanhaoranCC** [reply]
```python
import torch
from torch_geometric.nn import GCNConv, SAGEConv, GATConv, GraphSAGE
from torch_geometric.datasets import PPI
from torch_geometric.loader import DataLoader

path = "./"
train_dataset = PPI(path, split='train')
val_dataset = PPI(path, split='val')
test_dataset = PPI(path, split='test')
train_data = Batch.from_data_list(train_dataset)
train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=1, shuffle=False)
class Net(torch.nn.Module):
    def __init__(self,p2):
        super().__init__()
        self.conv = GraphSAGE(in_channels=train_dataset.num_features, hidden_channels=256,
                              out_channels=256, num_layers=3, dropout=p2, act="prelu", jk="last",
                              aggr='lstm')
        self.bn = nn.BatchNorm1d(256)
        self.action = nn.PReLU()

    def forward(self, x, edge_index):
        output = self.conv(x, edge_index)
        return self.action(output)
```
Hello, I changed the aggregation method to LSTM, he reported the following error, I would like to ask why this is, thank you very much for your answer.
```linux
  File "./ppi.py", line 56, in __init__
    aggr='lstm')
  File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python37\lib\site-packages\torch_geometric\nn\models\basic_gnn.py", line 78, in __init__
    self.init_conv(in_channels, hidden_channels, **kwargs))
  File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python37\lib\site-packages\torch_geometric\nn\models\basic_gnn.py", line 211, in init_conv
    return SAGEConv(in_channels, out_channels, **kwargs)
  File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python37\lib\site-packages\torch_geometric\nn\conv\sage_conv.py", line 52, in __init__
    super().__init__(**kwargs)
  File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python37\lib\site-packages\torch_geometric\nn\conv\message_passing.py", line 88, in __init__
    assert self.aggr in ['add', 'sum', 'mean', 'min', 'max', 'mul', None]
AssertionError
```

**Reply by rusty1s** [reply]
I think you need to install from master/nightly in order to utilize `LSTM`-style aggregation in `SAGEConv`, see [here](https://github.com/pyg-team/pytorch_geometric#nightly-and-master). It just got recently added.
